File: 316-remove-duplicate-letters/remove-duplicate-letters.go
func removeDuplicateLetters(s string) string {
    var counter = make(map[rune]int)

    for _, char := range s{
        counter[char] += 1
    }
    var stack = []rune{}
    inStack := 0

    for _, char := range s{
        counter[char]--
        bit := 1 << (char - 'a')
        
        if inStack & bit != 0{
            continue
        }

        for len(stack) > 0 && char < stack[len(stack) - 1] && counter[stack[len(stack) - 1]] > 0 {
            inStack ^= 1 << (stack[len(stack) - 1] - 'a')
            stack = stack[:len(stack) - 1]
        }
        stack = append(stack, char)
        inStack |= bit
    }

    var result []string

    for _, char := range stack{
        result = append(result, string(char))
    }

    return strings.Join(result, "")
}

File: 316-remove-duplicate-letters/remove-duplicate-letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:

        counter_word = Counter(s)
        in_stack = 0
        stack = []

        for char in s:
            counter_word[char] -= 1
            bit = 1 << (ord(char) - ord("a"))
            if in_stack & bit:
                continue
            while stack and char < stack[-1] and counter_word[stack[-1]] > 0:
                popped_char = stack.pop()
                in_stack ^= 1 << (ord(popped_char) - ord("a"))
                
            
            stack.append(char)
            in_stack |= bit
        
        return "".join(stack)
